rust/candid/src/types/subtype.rs

```diff
@@ -2,52 +2,66 @@ use super::internal::{find_type, Field, Label, Type};
 use crate::parser::typing::TypeEnv;
 use std::collections::{HashMap, HashSet};
 
+pub type Gamma = HashSet<(Type, Type)>;
+
 /// Check if t1 <: t2
-pub fn subtype(env: &TypeEnv, t1: &Type, t2: &Type) -> bool {
+pub fn subtype(gamma: &mut Gamma, env: &TypeEnv, t1: &Type, t2: &Type) -> bool {
     use Type::*;
     if t1 == t2 {
         return true;
     }
+    if matches!(t1, Var(_) | Knot(_)) || matches!(t2, Var(_) | Knot(_)) {
+        if !gamma.insert((t1.clone(), t2.clone())) {
+            return true;
+        }
+        let res = match (t1, t2) {
+            (Var(id), _) => subtype(gamma, env, env.rec_find_type(id).unwrap(), t2),
+            (_, Var(id)) => subtype(gamma, env, t1, env.rec_find_type(id).unwrap()),
+            (Knot(id), _) => subtype(gamma, env, &find_type(id).unwrap(), t2),
+            (_, Knot(id)) => subtype(gamma, env, t1, &find_type(id).unwrap()),
+            (_, _) => unreachable!(),
+        };
+        gamma.remove(&(t1.clone(), t2.clone()));
```

Comment:
I don't think you have to remove them: if you reach this point, the two types  _are_ subtypes, and you can use this information freely in further passes. Again, compare the ocaml code for subtyping in the Motoko compiler.

Reply:
We need to remove when `res` is `false`. The OCaml code doesn't need to do this because it's using a functional set. In Rust, `HashSet` is imperative.

Reply:
When `res` is false, the whole function exits, and you through away `gamma`, doesn’t it?